src/app/data-model/EncoderToTLV.h

```diff
@@ -0,0 +1,61 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/data-model/Encode.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLV.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/// Defines an abstract class of something that can be encoded
+/// into a TLV with a given data tag
+class EncoderToTLV
+{
+public:
+    virtual ~EncoderToTLV() = default;
+
+    virtual CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag) = 0;
+};
+
+/// An `EncoderToTLV` that uses `DataModel::Encode` to encode things.
+///
+/// Generally useful to encode things like <ClusterName>::Commands::<CommandName>::Type
+/// structures.
+template <typename T>
+class ObjectEncoderToTLV : public EncoderToTLV
```

Comment:
How about DataModelTypeEncodableToTLV?  Or ClusterObjectEncodableToTLV?

It's really unclear what "Object" on its own is doing there.  What sort of object?  Note that "value" here can totally be an integer or something....

Reply:
The namespace was frustrating: `DataModel::DataModel....` stutters.

I picked Object for now to not stutter and this is already inside DataModel.

Reply:
Picked `DirectEncodableToTLV` ... at least not `Object` anymore.

Reply:
Updated again to `EncodableType`